Add workflow to automatically check code style issues for PRs

TSRBerry opened the pull request:
This PR was originally planned to contain a lot more things, but formatting already took a while and I didn't expect our codebase to report so many issues (via dotnet-format).

A new workflow will now be triggered for every PR which will run dotnet-format to check for style issues.
If the workflow doesn't succeed the author of the PR will need to make some style changes, before we are able to merge the PR. The issues dotnet-format found will be uploaded as an artifact but also printed to the log of the workflow run.

This allows reviewers to focus more on actual code review instead of adding a lot of comments about minor styling issues. This will hopefully make reviews a little bit easier and authors still get feedback about code styling with the workflow.

I was able to automate silencing some of the issues dotnet-format reported with this [script](https://gist.github.com/TSRBerry/5460e3eaf8ca259211e08742ee10b08d).

---

Since this PR required a few other code changes to make dotnet-format happy, feel free to suggest improvements/changes to those as well.

Feedback about this approach would also be appreciated!

---

List of code style/quality rules we need to discuss:

- [x] [IDE0044](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0044): Add readonly modifier
- [ ] [IDE0051](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0051): Remove unused private member
- [ ] [IDE0052](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0052): Remove unread private member
- [ ] [IDE0059](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0059): Remove unnecessary value assignment
- [ ] [IDE0060](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0060): Remove unused parameter
- [x] [IDE0066](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0066): Use switch expression
- [x] [IDE1006](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/naming-rules#rule-id-ide1006-naming-rule-violation):  Naming rule violation

- [ ] [CS0169](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0169): The private field 'class member' is never used
- [ ] [CS0414](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0414): The private field 'field' is assigned but its value is never used
- [ ] [CS0649](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0649): Field 'field' is never assigned to, and will always have its default value 'value'
- [x] [CA1822](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1822): Mark members as static
- ~~[CA1834](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1834): Use StringBuilder.Append(char) for single character strings~~

gdkchan commented:
> All Ryujinx projects including the solution file were moved into a new src directory, since @gdkchan wanted to do that for a while now and I thought this PR might be a good opportunity to do it.

What does this mean for existing PRs?

TSRBerry commented:
If they rebase, they shouldn't have any issues. If they don't it depends on how Github deals with it.
I was able to rebase a few minutes ago and it already moved the files for me, I just had to add them.

TSRBerry commented:
I'd also offer help with that if anyone runs into issues.

gdkchan commented on src/ARMeilleure/Decoders/OpCodeT32MovImm16.cs:
```diff
@@ -4,7 +4,7 @@ class OpCodeT32MovImm16 : OpCodeT32Alu, IOpCode32AluImm16
         public int Immediate { get; }
 
-        public bool IsRotated => false;
+        public static bool IsRotated => false;
```
This is a rather weird change. It seems that the property is unused (which is what makes this change possible without breaking code), so it could just be removed.

TSRBerry replied:
Oh you are right. I started cleaning things up with automated fixes from dotnet-format itself, but that probably also did some of these changes.

gdkchan commented on src/ARMeilleure/Instructions/InstEmitHelper.cs:
```diff
@@ -100,78 +100,59 @@ public static int GetRegisterAlias(Aarch32Mode mode, int regIndex)
-                default: throw new ArgumentOutOfRangeException(nameof(regIndex));
-            }
+                8 => mode == Aarch32Mode.Fiq
+                                    ? RegisterAlias.R8Fiq
```
The indentation looks weird here.

gdkchan commented on src/ARMeilleure/Instructions/InstEmitSimdCvt32.cs:
```diff
@@ -225,24 +225,14 @@ private static Operand EmitRoundMathCall(ArmEmitterContext context, MidpointRoun
-                default:
-                    throw new ArgumentOutOfRangeException(nameof(rm));
-            }
+            var roundMode = rm switch
```
```suggestion
            return rm switch
```
Can't it just do this directly?

gdkchan commented on src/ARMeilleure/Instructions/InstEmitSimdCvt32.cs:
```diff
@@ -267,7 +257,9 @@ public static void Vcvt_RM(ArmEmitterContext context)
                         0b10 => Intrinsic.Arm64FcvtpuS,
                         0b11 => Intrinsic.Arm64FcvtmuS,
-                        _ => throw new ArgumentOutOfRangeException(nameof(rm))
+#pragma warning disable CA2208
```
Why do we need to disable this warning here? I tried reading the description for this warning but it's not clear what it doesn't like on this code.

TSRBerry replied:
> Maybe that should be using a different exception since the invalid value is not from a method parameter.

I had a list of exceptions open while looking at these warnings, but wasn't sure which one would be more fitting, do you have something in mind?

> BTW while looking at this, I noticed that the { for the switch expression is at the same line as the switch, which doesn't match what we do for most other switch expressions I believe.

True! I'll need to fix that tomorrow.

gdkchan replied:
> I had a list of exceptions open while looking at these warnings, but wasn't sure which one would be more fitting, do you have something in mind?

I'm not sure, I think when the invalid values comes from internal state, `InvalidOperationException` should be used.

gdkchan commented on src/ARMeilleure/IntermediateRepresentation/OperandType.cs:
```diff
@@ -0,0 +1,62 @@
+using System;
```
Weird that GitHub thinks this is a new file.

TSRBerry replied:
Yeah, probably too many changes in here for it to be recognized as a moved file.

(It should still show up as moved if you look at the commit where I moved everything directly)

gdkchan reviewed (commented):
I took a quick look, didn't look at everything yet since the diff is pretty large.

It might be easier to review if the change moving everything to a `src` sub-directory is done on a separate PR. GitHub is currently showing this:
> The diff you're trying to view is too large. We only load the first 3000 changed files. 

So I guess it is not showing all the changes.

About the style changes, the only one I'm not sure about is the `new()` change. I'm not sure if we should be enforcing it everywhere, it sounds like something that would come up quite often. There are a few other changes that I would probably take a while to get used to, but I don't think they would be triggered often enough to annoy me.

There are a few warnings that seems to be disabled with pragmas quite often (like IDE0066). I didn't actually look at the number of times we need to disable it, but if it happens too often, maybe it's better to just set `csharp_style_prefer_switch_expression = false` on the editorconfig to avoid needing to disable it many times.

gdkchan commented:
> If they rebase, they shouldn't have any issues. If they don't it depends on how Github deals with it. I was able to rebase a few minutes ago and it already moved the files for me, I just had to add them.

If all it takes is a rebase and GitHub will move the files automatically, it should be fine. I was worried that it would require manually moving the files that changed which would be a major annoyance.

gdkchan commented:
> Not sure if Github assists with this as well, but git using the command line worked very well for me.

Sorry, I meant "git" (the command line tool) rather than "GitHub". I edited my comment to correct this.

TSRBerry commented on src/Directory.Packages.props:
```diff
@@ -20,7 +20,7 @@
     <PackageVersion Include="LibHac" Version="0.18.0" />
     <PackageVersion Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.4" />
-    <PackageVersion Include="Microsoft.CodeAnalysis.CSharp" Version="4.5.0" />
+    <PackageVersion Include="Microsoft.CodeAnalysis.CSharp" Version="4.4.0" />
```
See this commit and the linked issues: https://github.com/Ryujinx/Ryujinx/pull/4670/commits/7a49dfa51b03b3ce581b57eed7d565e7e21d5bb6

dotnet-format seemed to behave strangely (and crashed when trying the analyzers run) until the version was downgraded to 4.4.0 again.

TSRBerry replied:
I updated to 4.6.0 with 43decfa54945554778118474153654fe83962c98, since the issues linked in the previous commit have all been resolved with the release of this version.

TSRBerry replied:
Hmm, CI still seems to struggle with that. I'll revert to `4.4.0` again.

I opened an issue about this here: https://github.com/dotnet/format/issues/1904

TSRBerry commented:
> About the style changes, the only one I'm not sure about is the new() change. I'm not sure if we should be enforcing it everywhere, it sounds like something that would come up quite often. There are a few other changes that I would probably take a while to get used to, but I don't think they would be triggered often enough to annoy me.
> 
> There are a few warnings that seems to be disabled with pragmas quite often (like IDE0066). I didn't actually look at the number of times we need to disable it, but if it happens too often, maybe it's better to just set csharp_style_prefer_switch_expression = false on the editorconfig to avoid needing to disable it many times.

My intention was to enforce the existing rules from editorconfig, but also consider how we address styling in PRs currently. If I made changes here which don't reflect the code style of Ryujinx, I should revert them and adapt editorconfig to match the code style instead.

TSRBerry commented:
> It might be easier to review if the change moving everything to a src sub-directory is done on a separate PR. GitHub is currently showing this:
> 
>     The diff you're trying to view is too large. We only load the first 3000 changed files.
> 
> So I guess it is not showing all the changes.

Does selecting everything after the commit where I moved the files work around this issue?

gdkchan commented:
> My intention was to enforce the existing rules from editorconfig, but also consider how we address styling in PRs currently. If I made changes here which don't reflect the code style of Ryujinx, I should revert them and adapt editorconfig to match the code style instead.

I was talking about cases where `#pragma` is being used to disable warnings around specific code. If we need to do that, then we are not enforcing the style in all cases. I think if we need to do that often,  it makes more sense to just change the editorconfig and disable the code style rule, rather than littering the code with `#pragma`s.

About the switch expression change, I like switch expressions, but I'm not a fan of switch expression with `or` patterns, especially the automatic fix that seems to make rather large chains of `or`s in the same line which makes it quite long and ugly. It would be nice if we could only enable it for the trivial cases, but I don't think this is possible.

gdkchan commented:
Unrelated to the discussion above, but it would be nice if we could catch naming violations too. For example, I can write a method like `public static void DO_SOMETHING()` which does not match the naming we use for methods at all, but it would not flag that as invalid. I think there is an analyser that can detect that, but there's no automatic fix.

TSRBerry commented:
> I was talking about cases where #pragma is being used to disable warnings around specific code. If we need to do that, then we are not enforcing the style in all cases. I think if we need to do that often, it makes more sense to just change the editorconfig and disable the code style rule, rather than littering the code with #pragmas.

Ah I misunderstood then! I didn't want to change too many things at once, so I added the #pragmas first, so we could discuss whether we want to disable these rules or enforce them properly.

(I should have kept track of them and added a list here, oh well.)

> About the switch expression change, I like switch expressions, but I'm not a fan of switch expression with or patterns, especially the automatic fix that seems to make rather large chains of ors in the same line which makes it quite long and ugly. It would be nice if we could only enable it for the trivial cases, but I don't think this is possible.

Hmm, I'll have to think about this some more, but it doesn't sound impossible to me, since analyzers should be able to detect logical patterns. We might be able to add an analyzer to this.

> I think there is an analyser that can detect that, but there's no automatic fix.

It's surprising there isn't but we might as well add one. I think it could save us a lot of work too.

TSRBerry commented:
> Maybe a bit overkill to have our own analyzer just to enforce switch expressions to be used in some cases.

It sounds fun to implement (currently), so I might add something to `Ryujinx.CustomTasks` if you don't mind! :D

Thealexbarney commented:
I think putting a comment on each `pragma disable` saying what the disabled warning is could make things a lot clearer

gdkchan commented:
It would be nice to have whitespace formatting too. We would have to disable it on places where it's worth keeping the alignment, such as tables, or table-like calls such as the one on `ARMeilleure/Decoders/OpCodeTable.cs`.

TSRBerry commented:
> I think putting a comment on each `pragma disable` saying what the disabled warning is could make things a lot clearer

Yeah, that's a good idea! But before I add this, could we discuss which `pragma disable` lines we want to keep and which need to be removed (and with that the editorconfig rules that need to be adjusted)?

> It would be nice to have whitespace formatting too. We would have to disable it on places where it's worth keeping the alignment, such as tables, or table-like calls such as the one on ARMeilleure/Decoders/OpCodeTable.cs.

Definitely! I'll need to look through the whitespace formatter to figure out if there already is a way to disable it for some sections, but if there isn't I'll add one.

TSRBerry commented:
> Definitely! I'll need to look through the whitespace formatter to figure out if there already is a way to disable it for some sections, but if there isn't I'll add one.

I'm pretty sure disabling should work exactly the same way as it does for the other formatters, so I should be able to add `#pragma warning disable IDE0055` for exclusions.

But it will take me a while to fix all the current issues, since `dotnet format whitespace` currently reports 3249 files with issues.
Here is the current report.json it generates: [report.zip](https://github.com/Ryujinx/Ryujinx/files/11248680/report.zip) (79.1 MB)

gdkchan commented:
> I'm pretty sure disabling should work exactly the same way as it does for the other formatters, so I should be able to add `#pragma warning disable IDE0055` for exclusions.
> 
> But it will take me a while to fix all the current issues, since `dotnet format whitespace` currently reports 3249 files with issues. Here is the current report.json it generates: [report.zip](https://github.com/Ryujinx/Ryujinx/files/11248680/report.zip) (79.1 MB)

Personally I wouldn't worry too much about keeping superfluous alignment. As I said before, the only ones I think are worth keeping, are for tables that would be much harder to read without the alignment. For things like variable, field, proper declarations or method parameters, I don't think they are adding a lot of value.

For example:
```cs
Mammal  cat    = default;
Mammal  dog    = default;
Reptile turtle = default;
```
I think it would be fine as:
```cs
Mammal cat = default;
Mammal dog = default;
Reptile turtle = default;
```

TSRBerry commented:
The remaining files dotnet-format is unhappy about are all the files containing manual alignments.

Sooo, if you have a few minutes/hours to kill feel free to help me in making dotnet-format happy here :p

I'll continue working on this tomorrow.

TSRBerry commented:
@gdkchan Could you take a look at 3f9e20218e9da11369a4ddffe20dace74f5d900d?
This commit contains all the whitespace formatting changes and also all the places where I stopped it from auto formatting.

Just to make sure it's fine this way.

If you have a few places where it should be reverted and auto formatting disabled and vice-versa please let me know.

TSRBerry commented:
Since the formatting changes were applied building Ryujinx now reports a lot more warnings than before.
I'm not sure why that is the case and if we also want to address that within this PR, although that might be way too hard to review.

<details><summary>Build warnings</summary>

```
Ryujinx-formatting/src/ARMeilleure/Statistics.cs(16,34): warning CS0169: The field 'Statistics._executionTimer' is never used [Ryujinx-formatting/src/ARMeilleure/ARMeilleure.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(33,21): warning CS0649: Field 'PictureInfo.ChromaTopFieldOffset' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(19,21): warning CS0649: Field 'PictureInfo.PicHeightInMbs' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(24,21): warning CS0649: Field 'PictureInfo.NumRefIdxL1ActiveMinus1' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(36,21): warning CS0649: Field 'PictureInfo.HistBufferSize' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(17,21): warning CS0649: Field 'PictureInfo.FrameMbsOnlyFlag' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(22,21): warning CS0649: Field 'PictureInfo.PicOrderPresentFlag' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(40,38): warning CS0649: Field 'PictureInfo.ScalingLists4x4' is never assigned to, and will always have its default value [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(23,21): warning CS0649: Field 'PictureInfo.NumRefIdxL0ActiveMinus1' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(21,21): warning CS0649: Field 'PictureInfo.EntropyCodingModeFlag' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(50,21): warning CS0649: Field 'PictureInfo.Flags2' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(44,23): warning CS0649: Field 'PictureInfo.Padding2A2' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(41,38): warning CS0649: Field 'PictureInfo.ScalingLists8x8' is never assigned to, and will always have its default value [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(31,21): warning CS0649: Field 'PictureInfo.LumaBottomFieldOffset' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(18,21): warning CS0649: Field 'PictureInfo.PicWidthInMbs' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(9,23): warning CS0169: The field 'PictureInfo.Unknown0' is never used [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(26,21): warning CS0649: Field 'PictureInfo.RedundantPicCntPresentFlag' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(27,21): warning CS0649: Field 'PictureInfo.Transform8x8ModeFlag' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(12,21): warning CS0649: Field 'PictureInfo.NumSlices' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(28,21): warning CS0649: Field 'PictureInfo.LumaPitch' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(38,28): warning CS0649: Field 'PictureInfo.FieldOrderCnt' is never assigned to, and will always have its default value [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(43,21): warning CS0649: Field 'PictureInfo.MvcextNumInterViewRefsL1' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(35,21): warning CS0649: Field 'PictureInfo.ChromaFrameOffset' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(46,21): warning CS0649: Field 'PictureInfo.Unknown2A8' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(16,21): warning CS0649: Field 'PictureInfo.DeltaPicOrderAlwaysZeroFlag' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(39,40): warning CS0649: Field 'PictureInfo.RefFrames' is never assigned to, and will always have its default value [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(47,21): warning CS0649: Field 'PictureInfo.Unknown2AC' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(20,21): warning CS0649: Field 'PictureInfo.BlockLayout' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(51,30): warning CS0649: Field 'PictureInfo.Unknown2D4' is never assigned to, and will always have its default value [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(32,21): warning CS0649: Field 'PictureInfo.LumaFrameOffset' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(11,21): warning CS0649: Field 'PictureInfo.BitstreamSize' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(42,21): warning CS0649: Field 'PictureInfo.MvcextNumInterViewRefsL0' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(37,22): warning CS0649: Field 'PictureInfo.Flags' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(45,21): warning CS0649: Field 'PictureInfo.Unknown2A4' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(14,21): warning CS0649: Field 'PictureInfo.Unknown54' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(15,21): warning CS0649: Field 'PictureInfo.Log2MaxPicOrderCntLsbMinus4' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(34,21): warning CS0649: Field 'PictureInfo.ChromaBottomFieldOffset' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(48,30): warning CS0649: Field 'PictureInfo.MvcextViewRefMasksL0' is never assigned to, and will always have its default value [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(30,21): warning CS0649: Field 'PictureInfo.LumaTopFieldOffset' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(49,30): warning CS0649: Field 'PictureInfo.MvcextViewRefMasksL1' is never assigned to, and will always have its default value [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(29,21): warning CS0649: Field 'PictureInfo.ChromaPitch' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(25,21): warning CS0649: Field 'PictureInfo.DeblockingFilterControlPresentFlag' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Types/H264/PictureInfo.cs(13,21): warning CS0649: Field 'PictureInfo.Unknown50' is never assigned to, and will always have its default value 0 [Ryujinx-formatting/src/Ryujinx.Graphics.Nvdec/Ryujinx.Graphics.Nvdec.csproj]
Ryujinx-formatting/src/Ryujinx.HLE/HOS/Services/Hid/Hid.cs(99,32): warning CS1696: Single-line comment or end-of-line expected [Ryujinx-formatting/src/Ryujinx.HLE/Ryujinx.HLE.csproj]
Ryujinx-formatting/src/Ryujinx.HLE/HOS/Services/Hid/Types/SharedMemory/Common/RingLifo.cs(13,32): warning CS0169: The field 'RingLifo<T>._unused' is never used [Ryujinx-formatting/src/Ryujinx.HLE/Ryujinx.HLE.csproj]
```
</details>

TSRBerry commented:
Forget the thing about the warnings, I looked at the history of these files and saw the warnings were silenced before.

TSRBerry commented:
Addressed some feedback I received over discord.

gdkchan commented:
> [IDE0044](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0044): Add readonly modifier
[IDE0051](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0051): Remove unused private member
[IDE0052](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0052): Remove unread private member
[IDE0059](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0059): Remove unnecessary value assignment

I think those are good, there's no point in keeping unused members. The `readonly` modifier can enable some optimizations I think (also prevents the value from being changed outside of the constructor which is generally good).

> [IDE0060](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0060): Remove unused parameter

This also seems good assuming it will not suggest removing unused parameters in places it can't be removed, like when it is required to implement an interface (I think I have already seen it suggesting that...).

> [IDE0066](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0066): Use switch expression

I have mixed feelings about this one because some of the suggestions doesn't look very good. On switch with multiple `case`s leading to the same path for example, it looks terrible. For the simple case where there's a single `case` followed by an assignment or return, I think it looks good.

For example, I think it looks bad in in this case:
```cs
// Good
switch (value)
{
    case Format.D24S8:
    case Format.D32:
    case Format.D24:
    case Format.D16:
        return true;
    default:
        return false:
}
// Bad
return value switch
{
    Format.D24S8 or Format.D32 or Format.D24 or Format.D16 => true,
    _ => false
};
```
Case where it looks good:
```cs
// Good
switch (value)
{
    case Format.D24S8:
        return 8;
    case Format.D32:
        return 4;
    case Format.D24:
        return 3;
    case Format.D16:
        return 2;
    default:
        return 0:
}
// Event better
return value switch
{
    Format.D24S8 => 8,
    Format.D32 => 4,
    Format.D24 => 3,
    Format.D16 => 2,
    _ => 0
};
```
> [IDE1006](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/naming-rules#rule-id-ide1006-naming-rule-violation): Naming rule violation

This is good, we should enforce that things are named properly.

> [CS0169](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0169): The private field 'class member' is never used

Good.

> [CS0414](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0414): The private field 'field' is assigned but its value is never used

This one is problematic, in some cases the compiler can't determine that fields are used because they are part of a blittable  struct that is reinterpreted as something else. We have already been adding pragmas for silence it when required though, so I think it's fine to keep it enabled.

> [CS0649](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0649): Field 'field' is never assigned to, and will always have its default value 'value'

Same as above.

> [CA1822](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1822): Mark members as static

Good, if something can be static, there's no reason to not make it so. It's also a minor optimization as static member don't need the `this` pointer to access the object instance.

TSRBerry commented:
Would you be in favor of resolving the current issues instead of silencing them?

gdkchan commented:
> Would you be in favor of resolving the current issues instead of silencing them?

Yes, I think resolving the issues if possible is better than adding pragmas, or disabling the suggestion.

TSRBerry commented:
> > [IDE0060](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0060): Remove unused parameter
> 
> This also seems good assuming it will not suggest removing unused parameters in places it can't be removed, like when it is required to implement an interface (I think I have already seen it suggesting that...).

We have a lot of cases in `Ryujinx.HLE` where we aren't using the context parameter, which produces a lot of these warnings. Should we also resolve all of these? (Maybe by using "_" as the parameter name?)

gdkchan commented:
> > > [IDE0060](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0060): Remove unused parameter
> > 
> > 
> > This also seems good assuming it will not suggest removing unused parameters in places it can't be removed, like when it is required to implement an interface (I think I have already seen it suggesting that...).
> 
> We have a lot of cases in `Ryujinx.HLE` where we aren't using the context parameter, which produces a lot of these warnings. Should we also resolve all of these? (Maybe by using "_" as the parameter name?)

If they are unused, I think it can be just removed? I don't think there's any point in keeping them unless I'm missing something.

TSRBerry commented:
> > [IDE1006](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/naming-rules#rule-id-ide1006-naming-rule-violation): Naming rule violation
> 
> This is good, we should enforce that things are named properly.

I think we need to have one exception for PInvoke related methods and variables, right?

TSRBerry commented:
> If they are unused, I think it can be just removed? I don't think there's any point in keeping them unless I'm missing something.

I always thought we can't remove them, but if we can then yeah there is no point in keeping them!

gdkchan commented:
> > > [IDE1006](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/naming-rules#rule-id-ide1006-naming-rule-violation): Naming rule violation
> > 
> > 
> > This is good, we should enforce that things are named properly.
> 
> I think we need to have one exception for PInvoke related methods and variables, right?

Can you clarify? For P/Invoke methods, we don't provide the implementation, the implementation is on the native library and we only provide the signature, so the compiler has no way to know which parameter might be unused there, so it shouldn't produce any warnings.

TSRBerry commented:
This might be a good example:
https://github.com/Ryujinx/Ryujinx/blob/18f2fa4cc3dc3b73f0c7526b5b6a3fb65a4b6e5e/src/ARMeilleure/Signal/UnixSignalHandlerRegistration.cs#L23-L34

gdkchan commented:
And what exactly is unused on that example?

TSRBerry commented:
There is nothing unused, but everything in that snippet is a `Naming rule violation`, which is why I think we should have an exception for P/Invoke related stuff.

gdkchan commented:
> There is nothing unused, but everything in that snippet is a `Naming rule violation`, which is why I think we should have an exception for P/Invoke related stuff.

I think it would be fine to adjust the constant names to have proper naming, but we can't change the method names since they must match the name that is exported on the native library (although I think there is a way to specify the native library function name separately as a string, and then use a different name on the method, but I don't think it's worth doing that just for this name violation).

TSRBerry commented:
Is there a rule I could set to enforce that catch-blocks need to specify an exception?
I'd like to be able to deny something like this:

```cs
try
{
    // Some code ....
}
catch
{
    // Some code that silently deals with this exception.
}
```

TSRBerry commented:
Rebased and dropped all the commits about moving the projects to `src`.
Hopefully this is also a bit easier to review now.

TSRBerry commented:
Rebased, all projects are now in the src directory and I didn't need to solve any conflicts, it just worked!

TSRBerry commented:
I would like to ask for your feedback on the rules as well:
- @marysaka 
- @AcK77 
- @riperiperi

TSRBerry commented:
@Ryujinx/developers I don't want to rebase this again, before we discussed the different code quality/style rules which are listed in the description. I'd like to hear opinions about the rules that should be changed as well as the pragmas we should keep or resolve.

TSRBerry commented:
In 1 or 2 weeks I'll rebase this again and start working on resolving the remaining issues, based on the feedback in here.

I hope this is enough time to get a little more feedback as well.

marco-carvalho commented:
Wouldn't be better to break this PR into smaller ones? 20k changes sounds tough to review.

TSRBerry commented:
Yes, that's what I tried to do already, but it's still a lot as you can see. I'm mostly looking for code-style rules feedback here, since almost all of these changes are automated.

I think it has to be done once, since this wasn't enforced as the project was created.

marysaka commented on .github/workflows/checks.yml:
```diff
@@ -0,0 +1,40 @@
+name: Perform checks
```
In the future it would be nice to make that the first job to run and not run any builds jobs if it fails to avoid uneeded CI resources usage.

TSRBerry replied:
Will configure that in here as well!

marysaka reviewed (approved):
I'm fine with the code-style changes even tho some are probably going to take some time to settle (but automated check help a lot!)

TSRBerry commented:
![image](https://github.com/Ryujinx/Ryujinx/assets/20988865/4f92871b-3e70-4d15-a974-746c84bd3141)

With the last commit, I won't be able to create new builds in this PR, but as long as the checks workflow runs that should be okay.

gdkchan commented on src/Ryujinx.Horizon/Ryujinx.Horizon.csproj.DotSettings:
```diff
@@ -0,0 +1,2 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Why are those files needed?

gdkchan commented on src/ARMeilleure/Decoders/OpCodeTable.cs:
```diff
@@ -1347,8 +1349,7 @@ private static void SetT16(string encoding, InstName name, InstEmitter emitter,
-            MakeOp reversedMakeOp =
-                (inst, address, opCode)
+            OpCode reversedMakeOp(InstDescriptor inst, ulong address, int opCode)
                     => makeOp(inst, address, (int)BitOperations.RotateRight((uint)opCode, 16));
```
Indentation seems a bit off here now.

gdkchan commented on .editorconfig:
```diff
@@ -158,46 +160,4 @@ csharp_space_between_square_brackets = false
-
-# Naming rules
-
-dotnet_naming_rule.interface_should_be_begins_with_i.severity = suggestion
```
Why all those naming rules were removed?

TSRBerry replied:
These are just the default ones, so they just clutter the config. It's better to only include the rules we changed.

(See: 6380c72a4faae6e644b7e5143b8ca29a127a5008)

TSRBerry replied:
I was wrong about this.
dotnet format doesn't seem to know about the default naming rules, so including them in our `.editorconfig` is actually better.

TSRBerry commented:
@Ryujinx/developers quick overview of the amount of disabled warnings we currently have:

| Project | # of 'disable warning' statements |
|------------|-------------------------------------------------:|
| **Total** | 942 |
| ARMeilleure | 35 |
| Ryujinx | 26 |
| Ryujinx.Audio | 18 |
| Ryujinx.Audio.Backends.SDL2 | 1 |
| Ryujinx.Ava | 9 | 
| Ryujinx.Common | 76 |
| Ryujinx.Cpu | 13 |
| Ryujinx.Graphics.Gpu | 68 |
| Ryujinx.Graphics.Host1x | 2 |
| Ryujinx.Graphics.Nvdec | 11 |
| Ryujinx.Graphics.Nvdec.FFmpeg | 7 |
| Ryujinx.Graphics.Nvdec.Vp9 | 13 |
| Ryujinx.Graphics.OpenGL | 4 |
| Ryujinx.Graphics.Shader | 43 |
| Ryujinx.Graphics.Texture | 3 |
| Ryujinx.Graphics.Vic | 8 |
| Ryujinx.Graphics.Vulkan | 19 |
| Ryujinx.Headless.SDL2 | 1 |
| Ryujinx.HLE | 523 |
| Ryujinx.Horizon | 18 |
| Ryujinx.Input | 1 |
| Ryujinx.Input.SDL2 | 1 |
| Ryujinx.Memory | 6 |
| Ryujinx.Tests | 30 |
| Ryujinx.Tests.Memory | 3 |
| Ryujinx.Tests.Unicorn | 1 |
| Ryujinx.Ui.Common | 1 |
| Spv.Generator | 1 |

How do we deal with these? Do we want to get rid of all (excluding formatting) of them?
Most of the ones in `Ryujinx.HLE` are for all the service methods which aren't using the service context.

If we want to remove most of these, could I get some help with some projects? Especially for `Ryujinx.Graphics*` and `Ryujinx.Audio*`, since I don't want to make any changes that would remove stuff you added for a reason, but isn't used yet.

TSRBerry commented:
Seems like I broke Astc with `4271d522d` somehow. I'll need to check that again.

marco-carvalho commented:
@TSRBerry Not trying to rain on your parade, but don't you think this PR is stalling other PRs (https://github.com/Ryujinx/Ryujinx/pull/5227 and https://github.com/Ryujinx/Ryujinx/pull/5215) that could already be merged, but are now closed because you said this PR is already addressing those issues?

TSRBerry commented:
I agree with you actually. Sorry about that, I thought we could get this in a little faster tbh.
Maybe these can be reopened and merged before and I'll just rebase this PR again.

TSRBerry commented:
Since dotnet/format#1800 has been resolved I'll try updating Microsoft.CodeAnalysis.CSharp to 4.6.0 again.

marco-carvalho commented:
> I agree with you actually. Sorry about that, I thought we could get this in a little faster tbh. Maybe these can be reopened and merged before and I'll just rebase this PR again.

Don't need to say sorry, bro. Honestly I didn't saw that you already resolved those code smells in this PR when I opened https://github.com/Ryujinx/Ryujinx/pull/5227 and https://github.com/Ryujinx/Ryujinx/pull/5215, so the closing made sense.
@marysaka could you reopen them, please?

TSRBerry commented:
This PR contains the changes to `.editorconfig` and the GitHub workflows. The version adjustment for the CodeAnalysis dependency might also be removed if everything works with the latest version as well!

I'll create the PRs for every project now!

TSRBerry commented:
I split my commits using this python script I wrote: https://gist.github.com/TSRBerry/b0ffbd031b0366e6cf8b6ecf50359a0e

IsaacMarovitz commented:
<img width="848" alt="Screenshot 2023-06-24 at 21 35 57" src="https://github.com/Ryujinx/Ryujinx/assets/42140194/9ca7853b-354a-4095-a91e-2b461877d469">

Probably my biggest qualm with these changes is that it messes up bits that were intentionally lined up in a non-standard way. I feel these can really improve code readability in repetitive sections, and I would prefer to see them stay.

TSRBerry commented:
See [this comment](https://github.com/Ryujinx/Ryujinx/pull/4670#issuecomment-1510563750) and the replies. I disabled formatting for all the table like alignments in the repo. Would that be sufficient?

IsaacMarovitz commented:
> See [this comment](https://github.com/Ryujinx/Ryujinx/pull/4670#issuecomment-1510563750) and the replies. I disabled formatting for all the table like alignments in the repo. Would that be sufficient?

Okay, I'll comment on the specific place where the one I found occurs

marysaka reviewed (approved):
only one comment otherwise lgtm

TSRBerry commented on .github/workflows/nightly_pr_comment.yml:
```diff
@@ -1,8 +1,10 @@
+
 on:
   workflow_run:
-    workflows: ['Build job']
+    workflows: ['Perform checks']
```
See: https://github.com/Ryujinx/Ryujinx/actions/runs/5586425962

The checks workflow now calls the build workflow, so it doesn't run as a standalone workflow anymore.
Although it's never explicitly stated anywhere, I'm thinking the comment workflow won't be triggered anymore if the build workflow is not running standalone anymore. And since the checks workflow only finishes if the builds were created or dotnet format found issues it should work fine with this change.

gdkchan commented on src/Ryujinx.Graphics.OpenGL/Queries/CounterQueue.cs:
```diff
@@ -13,7 +13,9 @@ class CounterQueue : IDisposable
         public CounterType Type { get; }
         public bool Disposed { get; private set; }
 
+#pragma warning disable IDE0052 // Private member is never read
```
Remove it instead of adding this pragma?

gdkchan commented on .editorconfig:
```diff
@@ -1,8 +1,7 @@
 
-# C# files
-[*.cs]
+[*]
```
I guess those rules will apply to all file extensions, is that fine? I guess it wouldn't apply to files it doesn't make sense to right?

TSRBerry replied:
At least currently, we don't have any files where it wouldn't make sense to have this imo.
